ext/objspace/objspace_dump.c

```diff
@@ -25,34 +25,127 @@
 static VALUE sym_output, sym_stdout, sym_string, sym_file;
 static VALUE sym_full;
 
+static const unsigned long buffer_capacity = 4096;
+
 struct dump_config {
     VALUE type;
-    FILE *stream;
+    VALUE stream;
     VALUE string;
     const char *root_category;
     VALUE cur_obj;
     VALUE cur_obj_klass;
     size_t cur_obj_references;
     unsigned int roots: 1;
     unsigned int full_heap: 1;
+    unsigned long buffer_len;
+    char buffer[buffer_capacity];
 };
 
-PRINTF_ARGS(static void dump_append(struct dump_config *, const char *, ...), 2, 3);
 static void
-dump_append(struct dump_config *dc, const char *format, ...)
+dump_flush(struct dump_config *dc)
+{
+    if (dc->buffer_len) {
+        if (dc->stream) {
+            size_t written = rb_io_bufwrite(dc->stream, dc->buffer, dc->buffer_len);
+            if (written < dc->buffer_len) {
+                MEMMOVE(dc->buffer, dc->buffer + written, char, dc->buffer_len - written);
+                dc->buffer_len -= written;
+                return;
+            }
+        }
+        else if (dc->string) {
+            rb_str_cat(dc->string, dc->buffer, dc->buffer_len);
+        }
+        dc->buffer_len = 0;
+    }
+}
+
+static inline void
+buffer_ensure_capa(struct dump_config *dc, unsigned long requested)
+{
+    RUBY_ASSERT(requested <= buffer_capacity);
+    if (requested + dc->buffer_len >= buffer_capacity) {
+        dump_flush(dc);
+        if (requested + dc->buffer_len >= buffer_capacity) {
+            // TODO: raise?
```

Comment:
Sounds like my comment was lost somehow:

Not sure what do do in this case (incomplete fwrite and full buffer). I think it is safe to raise, but not sure what error would be best to raise.